Array speedup

tautschnig opened the pull request.

kroening committed:
performance tests for reading/writing arrays

kroening committed:
Array theory: implement weakly equivalent arrays

This implements Christ and Hoenicke's Weakly Equivalent Arrays
(https://arxiv.org/pdf/1405.6939.pdf) with in-place depth-first path
enumeration.

Co-authored-by: Michael Tautschnig <tautschn@amazon.com>

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -8,6 +8,9 @@ Author: Daniel Kroening, kroening@kroening.com
 
+#include <cassert>
+#include <iostream>
+
```
Don't include these. All outputs should be in appropriate mstreamt and assertions should be replaced by invariants and such.

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -49,47 +43,39 @@ literalt arrayst::record_array_equality(
-      false,
-      "record_array_equality got equality without matching types");
+    std::cout << equality.pretty() << '\n';
+    throw "record_array_equality got equality without matching types";
```
use INVARIANT

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -49,47 +43,39 @@ literalt arrayst::record_array_equality(
+            "equality has array type");
+
+  std::size_t a1=arrays.number(op0);
+  std::size_t a2=arrays.number(op1);
```
const ?

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -49,47 +43,39 @@ literalt arrayst::record_array_equality(
   {
-    forall_operands(op, expr) collect_indices(*op);
+    forall_operands(op, expr)
+      collect_indices(*op);
```
I would suggest `std::for_each(expr.operands().begin(), expr.operands().end(), collect_indices)`

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -111,23 +95,29 @@ void arrayst::collect_arrays(const exprt &a)
     to_array_type(ns.follow(a.type()));
 
+  // remember it
+  wegt::node_indext a_ind=arrays.number(a);
```
`a_index` (and maybe const)

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -111,23 +95,29 @@ void arrayst::collect_arrays(const exprt &a)
-      prop.error() << a.pretty() << messaget::eom;
-      DATA_INVARIANT(false, "collect_arrays got 'with' without matching types");
+      std::cout << a.pretty() << '\n';
+      throw "collect_arrays got 'with' without matching types";
```
INVARIANT

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -99,9 +85,7 @@ void arrayst::collect_indices(const exprt &expr)
       if(is_unbounded_array(array_type))
-      {
         record_array_index(e);
-      }
```
Could be part of a separate reformatting commit.

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -111,23 +95,29 @@ void arrayst::collect_arrays(const exprt &a)
+    collect_indices(with_expr.new_value());
+
+    // one undirected edge for each array update
+    std::size_t expr_old_ind=arrays.number(with_expr.old());
```
`expr_old_index` (const?)

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -136,20 +126,14 @@ void arrayst::collect_arrays(const exprt &a)
-        false,
-        "collect_arrays got 'update' without matching types");
+      std::cout << a.pretty() << '\n';
+      throw "collect_arrays got 'update' without matching types";
```
INVARIANT

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -158,19 +142,24 @@ void arrayst::collect_arrays(const exprt &a)
-      prop.error() << a.pretty() << messaget::eom;
-      DATA_INVARIANT(false, "collect_arrays got if without matching types");
+      std::cout << a.pretty() << '\n';
+      throw "collect_arrays got if without matching types";
```
INVARIANT

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -158,19 +142,24 @@ void arrayst::collect_arrays(const exprt &a)
-      prop.error() << a.pretty() << messaget::eom;
-      DATA_INVARIANT(false, "collect_arrays got if without matching types");
+      std::cout << a.pretty() << '\n';
+      throw "collect_arrays got if without matching types";
```
INVARIANT

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -158,19 +142,24 @@ void arrayst::collect_arrays(const exprt &a)
 
-    arrays.make_union(a, if_expr.true_case());
-    arrays.make_union(a, if_expr.false_case());
+    // do weg edges
```
This comment does not add information. And what does weg mean?

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -158,19 +142,24 @@ void arrayst::collect_arrays(const exprt &a)
-    arrays.make_union(a, if_expr.false_case());
+    // do weg edges
+    std::size_t expr_true_ind=arrays.number(if_expr.true_case());
+    std::size_t expr_false_ind=arrays.number(if_expr.false_case());
```
const, index

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -182,10 +171,9 @@ void arrayst::collect_arrays(const exprt &a)
-       a.op0().id_string()+"'").c_str());
+    if(to_member_expr(a).struct_op().id()!=ID_symbol)
+      throw
+        "unexpected array expression: member with `"+a.op0().id_string()+"'";
```
keep the DATA_INVARIANT

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
-    DATA_INVARIANT(
-      false,
-      "byte_update should be removed before collect_arrays");
+    assert(false);
```
don't use assert

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
+      collect_arrays(op);
+    }
+    else
+      throw "unexpected array type cast from "+op.type().id_string();
```
INVARIANT

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
+    collect_arrays(to_index_expr(a).array());
   }
   else
+    throw "unexpected array expression (collect_arrays): `"+a.id_string()+"'";
```
UNREACHABLE

romainbrenguier commented on src/solvers/flattening/arrays.h:
```diff
@@ -48,81 +49,97 @@ class arrayst:public equalityt
-  {
-    lazy_typet type;
-    exprt lazy;
+  bool must_be_different(const exprt &, const exprt &);
```
const?

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
+  return false;
+}
+
+void arrayst::weg_path_condition(
```
should be documented

romainbrenguier replied:
using a verb in the function name is generally clearer

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
-       a.id_string()+"'").c_str());
+    std::cout << "edge: " << pn.n << "->" <<
+      pn.edge->first << " " <<
+      from_expr(arrays[pn.n]) << "\n";
```
don't write on `std::cout`

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
+    if(weg_edge.update.is_not_nil())
+    {
+      const auto &with_expr=to_with_expr(weg_edge.update);
+      notequal_exprt inequality(with_expr.where(), index_a);
```
const

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
-  if(lazy_arrays && refine)
+  INVARIANT(!path.empty(), "path is not empty");
+  wegt::node_indext a=path.front().n;
+  wegt::node_indext b=path.back().n;
```
const ?

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
+void arrayst::process_weg_path(const weg_patht &path)
 {
-  if(lazy_arrays && refine)
+  INVARIANT(!path.empty(), "path is not empty");
```
PRECONDITION ?

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
+
+  std::cout << "b is: "
+            << from_expr(ns, "", arrays[b])
+            << '\n';
```
don't write on std::cout

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
-    // lazily add the constraint
-    if(incremental_cache)
+    // lazily instantiate read-over-write
+    if(arrays[b].id()==ID_with)
```
use expr_try_dynamic_cast

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
+      weg_path_condition(path, index_a, cond);
+
+      // a_i=b_i
+      index_exprt a_i(arrays[a], index_a);
```
const?

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
+      // a_i=b_i
+      index_exprt a_i(arrays[a], index_a);
+      // cond => a_i=b_i
+      implies_exprt implication(
```
const?

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
+      implies_exprt implication(
+        conjunction(cond),
+        equal_exprt(a_i, value_b));
+      std::cout << "C2a: "
```
don't write on std::cout

romainbrenguier commented on src/solvers/flattening/arrays.cpp:
```diff
@@ -198,95 +186,213 @@ void arrayst::collect_arrays(const exprt &a)
+      set_to_true(implication);
     }
-    else
+    else if(arrays[b].id()==ID_array_of)
```
use expr_try_dynamic_cast

romainbrenguier reviewed (dismissed):
I only partially reviewed, but several issues need to be addressed.

martin-cs commented:
@tautschnig shout when you want a review from me.

TGWDB commented:
@tautschnig This PR is still waiting on some changes to address reviewer comments (and now a rebase). Is there a plan to progress on this PR since it is mentioned in others?

tautschnig commented:
> @tautschnig This PR is still waiting on some changes to address reviewer comments (and now a rebase). Is there a plan to progress on this PR since it is mentioned in others?

Yes, this is on my pile of TODOs. In terms of performance, it's probably one of the most urgent to work on. Just trying to address the soundness-related ones first.

martin-cs commented:
@tautschnig : Trevor and I have been having some issues with the current array code and may need to look at it in the next few months.  Reviewing this / getting it merged might well help us with that.

tautschnig committed:
Test trace-values: create reads to enforce instantiation

The test specification expects that the indices 0, 1, and one other are
instantiated. The array theory is only required to do so when also
reading from these elements.

tautschnig committed:
Use instantiate() when flattening array_comprehension_exprt

Uses the capabilities of binding_exprt instead of relying on the
unrelated replace_exprt to do the right thing.

tautschnig committed:
Flattening: support index expressions over zero bitwidth arrays

There is not really anything wrong in having empty bitvectors, which we
otherwise already support (as of e021eef4).

tautschnig committed:
Failing test to demonstrate need for extensionality rule

All existing tests rely on indexed access to arrays, which is covered by
the read-over-write axiom.

TGWDB commented:
Closing due to age (no further comment on PR content), please reopen with rebase on develop if you intent to continue this work.
